tests/test_denormalisation.py

```diff
@@ -0,0 +1,374 @@
+                    denormalised.denormalised_properties
+                )
+                assert model_class.parent.related_model == owner.__name__
+                assert (
+                    model_class.parent.remote_field.name ==
+                    '_denormalised_{}'.format(descriptor_name)
+                )
+
+                for prop in denormalised.denormalised_properties:
+                    assert getattr(model_class, prop.name) == prop.field
+
+                assert getattr(owner, descriptor_name) == denormalised
+
+        @mark.usefixtures('property')
+        def test_it_imports_the_denormalisation_model(
+            self, denormalised, owner, descriptor_name, patched_import_module
+        ):
+            denormalised.contribute_to_class(owner, descriptor_name)
+
+            patched_import_module.assert_called_once_with(owner.__module__)
+
+            assert (
+                getattr(
+                    patched_import_module(),
+                    denormalised.model_class.__name__
+                ) ==
+                denormalised.model_class
+            )
+
+        @mark.usefixtures('property')
+        def test_it_connects_all_signals(
+            self, denormalised, owner, descriptor_name, signals
+        ):
+            denormalised.contribute_to_class(owner, descriptor_name)
+
+            for signal in signals:
+                assert signal.connect.called
+
+        @mark.usefixtures('property')
+        def test_it_exposes_the_relation_on_the_instance(
+            self, denormalised, owner, descriptor_name, signals
+        ):
+            denormalised.contribute_to_class(owner, descriptor_name)
+
+            instance = create_autospec(
+                owner, instance=True,
+                **{
+                    denormalised.related_name: create_autospec(
+                        Manager, instance=True
+                    )
+                }
+            )
+
+            assert (
+                denormalised.__get__(instance, owner) ==
+                getattr(instance, denormalised.related_name)
+            )
+
+        class TestAttributesInitialisedOnSetName:
+            @mark.parametrize(
+                ('attr',),
+                (
+                    ('name',),
+                    ('model_class',),
+                    ('parent_model_class',),
+                )
+            )
+            def test_it_raises_an_attribute_error(self, denormalised, attr):
+                with raises(AttributeError):
+                    getattr(denormalised, attr)
+
+    @fixture
+    def denormalised_properties(self):
+        def foo(_self):
+            return 1
+
+        def bar(_self):
+            return 2
+
+        return (
+            DenormalisedProperty(foo, IntegerField(), ('other',)),
+            DenormalisedProperty(bar, IntegerField(), tuple()),
+        )
+
+    @fixture
+    def model_class(self, denormalised_properties):
+        return create_autospec(
```

Comment:
These tests are a little too dependent on mocks.  Ideally refactor to use `FooModel` etc.  (This is a little tricky due to the fact the descriptor dynamically defines new models.)

Reply:
No longer applies